swift/extractor/visitors/VisitorBase.h

```diff
@@ -21,14 +21,34 @@ class VisitorBase {
 
 // we want to override the default swift visitor behaviour of chaining calls to immediate
 // superclasses by default and instead provide our own TBD default (using the exact type)
-#define DEFAULT(KIND, CLASS, PARENT) \
-  void visit##CLASS##KIND(swift::CLASS##KIND* e) { dispatcher_.emitUnknown(e); }
+// if we the implementation class has translate##CLASS##KIND (that uses generated C++ classes), we
+// want to use that. We detect that by comparing the member function pointer to the one for a
+// private undefined member function of the same name and signature in {Ast,Type}VisitorBase, which
+// will be shadowed (and thus different) in case the implementation class has it.
+#define DEFAULT(KIND, CLASS, PARENT)                                              \
+ public:                                                                          \
+  void visit##CLASS##KIND(swift::CLASS##KIND* e) {                                \
+    constexpr bool hasTranslateImplementation =                                   \
+        (&Self::translate##CLASS##KIND != &CrtpSubclass::translate##CLASS##KIND); \
+    if constexpr (hasTranslateImplementation) {                                   \
+      TrapClassOf<swift::CLASS##KIND> entry{};                                    \
+      static_cast<CrtpSubclass*>(this)->translate##CLASS##KIND(e, &entry);        \
+      dispatcher_.emit(entry);                                                    \
+    } else {                                                                      \
+      dispatcher_.emitUnknown(e);                                                 \
+    }                                                                             \
+  }                                                                               \
+                                                                                  \
+ private:                                                                         \
+  void translate##CLASS##KIND(swift::CLASS##KIND*, TrapClassOf<swift::CLASS##KIND>*);
```

Comment:
Instead of taking an output parameter, can we use the return value for the TRAP class instead? See https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rf-out.

If not (presumably because it's also an input), we should document a contract: can it be null? any other requirements on it on entry? what happens to the input contents on success? on failure? etc. Should it be a reference instead of a pointer? Implementations seem to assume it's not null.

Reply:
good point. The pointer was mainly for consistency with the pointer to the swift entity, but thinking about it, as a new "visiting protocol" is provided here, we don't need that either. I've changed the translation signatures to a more semantically correct `codeql::XXXDecl translateXXX(const swift::XXXDecl& decl)`